File: frontend/src/app/api/portfolio.ts
import { apiFetch } from "./client";
import type {
  HoldingsResponse,
  PositionsResponse,
  MarginsResponse,
  PortfolioSummary,
  ApiHolding,
  ApiPosition,
} from "./types";
import type { Holding, Position, Margin } from "../data/mockData";

export function getHoldings(): Promise<HoldingsResponse> {
  return apiFetch<HoldingsResponse>("/portfolio/holdings");
}

export function getPositions(): Promise<PositionsResponse> {
  return apiFetch<PositionsResponse>("/portfolio/positions");
}

export function getMargins(): Promise<MarginsResponse> {
  return apiFetch<MarginsResponse>("/portfolio/margins");
}

export function getPortfolioSummary(): Promise<PortfolioSummary> {
  return apiFetch<PortfolioSummary>("/portfolio/summary");
}

// ---------------------------------------------------------------------------
// Mappers: snake_case API → camelCase frontend types
// ---------------------------------------------------------------------------

export function mapHolding(h: ApiHolding): Holding {
  return {
    symbol: h.tradingsymbol,
    exchange: h.exchange,
    quantity: h.quantity,
    t1Quantity: h.t1_quantity,
    avgPrice: h.average_price,
    ltp: h.last_price,
    dayChange: h.day_change,
    dayChangePercent: h.day_change_pct,
    pnl: h.pnl,
    pnlPercent: h.pnl_pct,
    currentValue: h.current_value,
    investedValue: h.invested_value,
    kpis: {},
  };
}

export function mapPosition(p: ApiPosition): Position {
  return {
    symbol: p.tradingsymbol,
    exchange: p.exchange,
    product: p.product,
    quantity: p.quantity,
    avgPrice: p.average_price,
    ltp: p.last_price,
    unrealisedPnl: p.unrealised,
    m2mPnl: p.pnl,
  };
}

export function mapMargins(m: MarginsResponse): Margin {
  return {
    available: m.equity.available_cash,
    used: m.equity.used_debits,
    total: m.equity.opening_balance,
  };
}

File: backend/schemas/portfolio.py
from __future__ import annotations

from pydantic import BaseModel


class Holding(BaseModel):
    tradingsymbol: str
    exchange: str
    instrument_token: int
    quantity: int
    t1_quantity: int
    average_price: float
    last_price: float
    close_price: float
    pnl: float
    pnl_pct: float
    day_change: float
    day_change_pct: float
    current_value: float
    invested_value: float


class HoldingsSummary(BaseModel):
    total_invested: float
    total_current_value: float
    total_pnl: float
    total_pnl_pct: float
    total_day_change: float
    total_day_change_pct: float


class HoldingsResponse(BaseModel):
    holdings: list[Holding]
    summary: HoldingsSummary


class Position(BaseModel):
    tradingsymbol: str
    exchange: str
    product: str
    quantity: int
    average_price: float
    last_price: float
    pnl: float
    unrealised: float
    realised: float


class PositionsResponse(BaseModel):
    positions: list[Position]


class EquityMargin(BaseModel):
    available_cash: float
    opening_balance: float
    used_debits: float


class MarginsResponse(BaseModel):
    equity: EquityMargin


class PortfolioSummary(BaseModel):
    total_invested: float
    current_value: float
    total_pnl: float
    total_pnl_pct: float
    available_margin: float
    holdings_count: int
    profitable_count: int
    loss_count: int
    xirr: float | None
